### PR TITLE
Re-add conf_template_source/conf_cookbook properties.

### PR DESCRIPTION
**Pull request by robbat2**
### Description

Re-add conf_template_source/conf_cookbook properties.

### Issues Resolved

https://github.com/sous-chefs/haproxy/issues/243

### Contribution Check List

- [ ] All tests pass.
- [X] New functionality includes testing.
- [ ] New functionality has been documented in the README if applicable

---
This change is [Reviewable](https://reviewable.io/reviews/sous-chefs/haproxy/244)

### CODE REVIEW

**Comment by robbat2** [comment]
Hmm, the accumulator pattern makes it really hard to keep source & cookbook set correctly. if I omit them, all of the `edit_resource` calls inherits the scope of that called a given resource (eg setting `cookbook` to `test`), even if you don't want that.

I don't see a `clean` solution yet, but trying to find some solution that reliably works.

**Comment by rshade** [comment]
probably the best way is to add resources for `peers` and `mailers`

**Comment by robbat2** [comment]
peers & mailers should be added yes, but they don't begin to enter the complexity of the rest of the configuration.

The main problem is that in the frontend & backend sections, the order of some directives is meaningful.
Here's a partial extract of my complex configuration, using frontend `http-request`/`acl`/`http-request` and backend `acl`/`http-response`

```
frontend foo
  http-request set-var(txn.origin) req.hdr(Origin)
  acl origin_panel var(txn.origin) -m end -i .panel.test1.example.com
  acl origin_panel var(txn.origin) -m sub -i .panel.test2.example.com
  acl origin_panel var(txn.origin) -m sub -i ...
  # Export fact as a boolean
  http-request set-var(txn.origin_is_panel) bool(true) if origin_panel
  http-request set-var(txn.origin_is_panel) bool(false) unless origin_panel

backend bar
  # Map boolean back to ACL
  acl origin_panel var(txn.origin_is_panel),bool
  acl res_status_403 status eq 403
  http-response set-status 200 if res_status_403 origin_panel METH_OPTIONS
  # etc

```

**Comment by rshade** [comment]
what happens with the same properties in each resource?

**Comment by robbat2** [comment]
If every resource had cookbook & template overridden, it would work, but feels crappy in code style / DRY-ness

**Comment by rshade** [comment]
Ok, good to know. IMO I would do it that way. There are a lot of repeated properties, I would not worry about the DRY part ATM.

**Comment by robbat2** [comment]
Yay, I figured out a way of DRY and make it pass tests!

File: resources/acl.rb [hunk 1]
```diff
@@ -8,8 +8,10 @@
   # As we're using the accumulator pattern we need to shove everything
   # into the root run context so each of the sections can find the parent
   with_run_context :root do
+    default_conf
     edit_resource(:template, config_file) do |new_resource|
-      cookbook 'haproxy'
+      source node['haproxy']['conf_template_source']
```
Comment (damacus) [inline resources/acl.rb]: I like what you're trying to do here buuuuut...

So we really don't want to be using node attributes here, as it messes with the resources "interface". It should be accessed only one way, and that's via the resource properties. 

The helper library you've added mirrors the `defaults` action of a resource :)

**Comment by damacus** [comment]
I think this approach is pretty sensislbe actualy. If we can't (and we probably don't want to!) offer every option thats available, and some of them won't fit into the extra options array, setting your own template or just throwing that file to disk yourself is probably even easier 👍

File: resources/acl.rb (hunk 1 above)
Comment (rshade) [inline resources/acl.rb]: Yeah I would rather it be less dry and use properties, then a library also.

Comment (robbat2) [inline resources/acl.rb]: Ok, I've moved away from node attributes (using `run_state` instead), and also dropped the library (repeated the default everywhere instead). Just waiting for the CI run to confirm it still works.

**Comment by robbat2** [comment]
@damacus @rshade please review again, fixes all of your concerns and passes tests.

**Comment by robbat2** [comment]
Reviewed 2 of 13 files at r2.
Review status: 2 of 17 files reviewed at latest revision, 1 unresolved discussion.

---

*[resources/acl.rb, line 13 at r1](https://reviewable.io:443/reviews/sous-chefs/haproxy/244#-Ko4enLiARvmCP3z8OpN-r1-13:-KoOyY7MM2ZfPyB5O3tu:b-896fix) ([raw file](https://github.com/sous-chefs/haproxy/blob/6371cfb8ed062e8d85b32e2b19e8023df143c8a8/resources/acl.rb#L13)):*
<details><summary><i>Previously, robbat2 (Robin H. Johnson) wrote…</i></summary><blockquote>

Ok, I've moved away from node attributes (using `run_state` instead), and also dropped the library (repeated the default everywhere instead). Just waiting for the CI run to confirm it still works.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/sous-chefs/haproxy/244)*

**Comment by damacus** [comment]
Forgive the possible n00b question here. How are you getting state into `node.run_state['haproxy']?

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/sous-chefs/haproxy/244#-:-KoP1LLbRtVs-Vakelhk:b4hsdws)*

**Comment by robbat2** [comment]
Ah, you probably missed this line in reviewing, where `run_state` gets populated with the user's input (or the default:
```
resources/install.rb:
 node.run_state['haproxy'] ||= { 'conf_template_source' => {}, 'conf_cookbook' => {} }
 node.run_state['haproxy']['conf_template_source'][config_file] = conf_template_source
 node.run_state['haproxy']['conf_cookbook'][config_file] = conf_cookbook
```

Then everywhere else that might instantiate the template, we have to ensure the default is passed in.
```
resources/listen.rb:      
  source lazy { node.run_state['haproxy']['conf_template_source'][config_file] ||= 'haproxy.cfg.erb' }
  cookbook lazy { node.run_state['haproxy']['conf_cookbook'][config_file] ||= 'haproxy' }
```

I think it might be doable without the `||=`, but safer not to assume otherwise at this point.

As to why the deep structure of `node.run_state['haproxy']['conf_template_source'][config_file]`; I realized that if there were multiple haproxy installs with different config_files and different templates, it would clobber in the state variable :-).

**Comment by damacus** [comment]
Yup! Completely missed it. Knew it was somewhere. Nice one.

**Comment by damacus** [comment]
Reviewed 5 of 18 files at r1, 12 of 13 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/sous-chefs/haproxy/244)*

**Comment by damacus** [comment]
@rshade I think this is ready to merge.

**Comment by rshade** [comment]
@robbat2 thank you very much for your PR. I know it has been a tough one. I am going to try and release this and the other open one at the same time.

**Pull request merged**